### PR TITLE
e2e: make-mirror

### PR DESCRIPTION
**Pull request by gyuho**
For https://github.com/coreos/etcd/issues/4589.

### CODE REVIEW

**Comment by gyuho** [comment]
/cc @heyitsanthony @xiang90 

And let me know if we have any missing e2e test.

Thanks!

File: e2e/etcd_test.go [hunk 1]
```diff
@@ -41,7 +40,8 @@ const (
 )
 
 var (
-	configNoTLS = etcdProcessClusterConfig{
+	etcdProcessBasePort = 20000
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: just make the base port a part of the `etcdProcessClusterConfig` and default to `etcdProcessBasePort` if not set?

File: e2e/ctl_v3_make_mirror_test.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	for i := range kvs {
+		if err = ctlV3Put(cx, kvs[i].key, kvs[i].val, ""); err != nil {
+			cx.t.Fatal(err)
+		}
+	}
+
+	if err = ctlV3Get(cx, []string{"key", "--prefix"}, kvs...); err != nil {
+		cx.t.Fatal(err)
+	}
+	if err = ctlV3Get(cx2, []string{"key", "--prefix"}, kvs...); err != nil {
```
Comment (heyitsanthony) [inline e2e/ctl_v3_make_mirror_test.go]: this seems racey-- how does it know the keys have been mirrored already?

File: e2e/etcd_test.go (hunk 1 above)
Comment (gyuho) [inline e2e/etcd_test.go]: sure will fix.

File: e2e/ctl_v3_make_mirror_test.go (hunk 2 above)
Comment (gyuho) [inline e2e/ctl_v3_make_mirror_test.go]: right...

1. retry logic
2. or sleep before `get`?

Do you have any suggestion?

Comment (heyitsanthony) [inline e2e/ctl_v3_make_mirror_test.go]: @gyuho `etcdctl watch`?

Comment (gyuho) [inline e2e/ctl_v3_make_mirror_test.go]: yeah right. good idea. will try that.

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Thanks!

File: e2e/ctl_v3_make_mirror_test.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+		if err != nil {
+			cx.t.Fatal(err)
+		}
+	}()
+
+	var (
+		kvs   = []kv{{"key1", "val1"}, {"key2", "val2"}, {"key3", "val3"}}
+		donec = make(chan struct{})
+	)
+	go func() {
```
Comment (heyitsanthony) [inline e2e/ctl_v3_make_mirror_test.go]: there's no need for a goroutine-- watch from `--rev=1` after the `ctlV3Get(cx, ...)`

File: e2e/etcd_test.go [hunk 4]
```diff
@@ -210,7 +211,10 @@ func (cfg *etcdProcessClusterConfig) etcdProcessConfigs() []*etcdProcessConfig {
 	for i := 0; i < cfg.clusterSize; i++ {
 		var curls []string
 		var curl, curltls string
-		port := etcdProcessBasePort + 2*i
+		if cfg.basePort == 0 {
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: move this check to the beginning of the method?

**Comment by gyuho** [comment]
@heyitsanthony Just fixed. PTAL. Thanks!

**Comment by heyitsanthony** [comment]
lgtm

**Comment by gyuho** [comment]
thanks. merging in greens.

**Pull request merged**